### ISSUE TITLE
failed: Error during WebSocket handshake: Unexpected response code: 200

### ISSUE THREAD

**Issue by rizkiaditya24** [miguelgrinberg/Flask-SocketIO#543]
Hi, I have no idea why my app with websocket works well when I test it on my local machine, but it won't connect when deploying on server. It keep displaying this error:

> WebSocket connnection to 'ws://xxx.xxx.xxx.xxx/socket.io/?EIO=3&transport=websocket&sid=sdfds9sd9879dsfsd0e0h783' failed: Error during WebSocket handshake: Unexpected response code: 200

I am guessing the problem is on my nginx config. I use eventlet by the way, here is my nginx conf:

```
upstream nodes {
  ip_hash;
  server 127.0.0.1:8000;
}

server {
  listen 80 default_server;
  listen [::]:80 default_server;
  listen 443 ssl;
  server_name api.example.com www.api.example.com;
  root /var/www/rest-api-v2;
  ssl_certificate /cert/bundle.crt;
  ssl_certificate_key /cert/my_key.key;

  location / {
      proxy_pass http://nodes;
      proxy_redirect off;
      proxy_set_header Host $host;
      proxy_set_header X-Real-IP $remote_addr;
      proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;

      if ($request_method = 'OPTIONS') {
        add_header 'Access-Control-Allow-Origin' 'http://localhost:3030' always;
        add_header 'Access-Control-Allow-Methods' 'GET, POST, PUT, DELETE, OPTIONS' always;
        add_header 'Access-Control-Allow-Credentials' 'true' always;
        add_header 'Access-Control-Allow-Headers' 'DNT,Authorization,X-CustomHeader,Keep-Alive,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Content-Range,Range' always;
        add_header 'Access-Control-Max-Age' 1728000 always;
        add_header 'Content-Type' 'text/plain; charset=utf-8' always;
        add_header 'Content-Length' 0 always;
        return 204;
      }

      if ($request_method = 'POST') {
        add_header 'Access-Control-Allow-Origin' 'http://localhost:3030' always;
        add_header 'Access-Control-Allow-Methods' 'GET, POST, PUT, DELETE, OPTIONS' always;
        add_header 'Access-Control-Allow-Credentials' 'true' always;
        add_header 'Access-Control-Allow-Headers' 'DNT,Authorization,X-CustomHeader,Keep-Alive,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Content-Range,Range' always;
        add_header 'Access-Control-Expose-Headers' 'DNT,X-CustomHeader,Keep-Alive,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Content-Range,Range' always;
      }

      if ($request_method = 'GET') {
        add_header 'Access-Control-Allow-Origin' 'http://localhost:3030' always;
        add_header 'Access-Control-Allow-Methods' 'GET, POST, PUT, DELETE, OPTIONS' always;
        add_header 'Access-Control-Allow-Credentials' 'true' always;
        add_header 'Access-Control-Allow-Headers' 'DNT,Authorization,X-CustomHeader,Keep-Alive,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Content-Range,Range' always;
        add_header 'Access-Control-Expose-Headers' 'DNT,X-CustomHeader,Keep-Alive,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Content-Range,Range' always;
      }

      if ($request_method = 'PUT') {
        add_header 'Access-Control-Allow-Origin' 'http://localhost:3030' always;
        add_header 'Access-Control-Allow-Methods' 'GET, POST, PUT, DELETE, OPTIONS' always;
        add_header 'Access-Control-Allow-Credentials' 'true' always;
        add_header 'Access-Control-Allow-Headers' 'DNT,Authorization,X-CustomHeader,Keep-Alive,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Content-Range,Range' always;
        add_header 'Access-Control-Expose-Headers' 'DNT,X-CustomHeader,Keep-Alive,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Content-Range,Range' always;
      }

      if ($request_method = 'DELETE') {
        add_header 'Access-Control-Allow-Origin' 'http://localhost:3030' always;
        add_header 'Access-Control-Allow-Methods' 'GET, POST, PUT, DELETE, OPTIONS' always;
        add_header 'Access-Control-Allow-Credentials' 'true' always;
        add_header 'Access-Control-Allow-Headers' 'DNT,Authorization,X-CustomHeader,Keep-Alive,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Content-Range,Range' always;
        add_header 'Access-Control-Expose-Headers' 'DNT,X-CustomHeader,Keep-Alive,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Content-Range,Range' always;
      }
  }
  location /socket.io {
    proxy_http_version 1.1;
    proxy_buffering off;
    proxy_redirect off;

    proxy_set_header Host $host;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;

    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection "Upgrade";
    proxy_pass http://nodes/socket.io;
  }
}
```

I've followed some solutions on the other similar issues, but nothing can solve my problem

**Comment by miguelgrinberg** [comment]
The nginx config looks okay, at least I don't see anything odd in there.

Can you try changing your eventlet server to listen on 0.0.0.0:8000 and then see if you can connect from the outside bypassing nginx?

**Comment by rizkiaditya24** [comment]
@miguelgrinberg thanks for your response. I've tried to run the eventlet with this command:

`gunicorn --worker-class eventlet -w 1 -b 0.0.0.0:8000 run:app`

But nothing change. I found that when I run the eventlet, it works well at first, the connection is established, but suddenly it will stop and give the error above on the client. I try to see the log, it gives me this error:

```
Traceback (most recent call last):
  File "/var/www/rest-api-v2/venv/lib/python3.5/site-packages/engineio/socket.py", line 30, in poll
    packets = [self.queue.get(timeout=self.server.ping_timeout)]
  File "/var/www/rest-api-v2/venv/lib/python3.5/site-packages/eventlet/queue.py", line 311, in get
    return waiter.wait()
  File "/var/www/rest-api-v2/venv/lib/python3.5/site-packages/eventlet/queue.py", line 140, in wait
    return get_hub().switch()
  File "/var/www/rest-api-v2/venv/lib/python3.5/site-packages/eventlet/hubs/hub.py", line 294, in switch
    return self.greenlet.switch()
queue.Empty

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/var/www/rest-api-v2/venv/lib/python3.5/site-packages/engineio/server.py", line 261, in handle_request
    environ, start_response)
  File "/var/www/rest-api-v2/venv/lib/python3.5/site-packages/engineio/socket.py", line 94, in handle_get_request
    raise e
  File "/var/www/rest-api-v2/venv/lib/python3.5/site-packages/engineio/socket.py", line 91, in handle_get_request
    packets = self.poll()
  File "/var/www/rest-api-v2/venv/lib/python3.5/site-packages/engineio/socket.py", line 33, in poll
    raise exceptions.QueueEmpty()
engineio.exceptions.QueueEmpty

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/var/www/rest-api-v2/venv/lib/python3.5/site-packages/gunicorn/workers/async.py", line 56, in handle
    self.handle_request(listener_name, req, client, addr)
  File "/var/www/rest-api-v2/venv/lib/python3.5/site-packages/gunicorn/workers/async.py", line 107, in handle_request
    respiter = self.wsgi(environ, resp.start_response)
  File "/var/www/rest-api-v2/venv/lib/python3.5/site-packages/flask/app.py", line 1994, in __call__
    return self.wsgi_app(environ, start_response)
  File "/var/www/rest-api-v2/venv/lib/python3.5/site-packages/flask_socketio/__init__.py", line 42, in __call__
    start_response)
  File "/var/www/rest-api-v2/venv/lib/python3.5/site-packages/engineio/middleware.py", line 47, in __call__
    return self.engineio_app.handle_request(environ, start_response)
  File "/var/www/rest-api-v2/venv/lib/python3.5/site-packages/socketio/server.py", line 353, in handle_request
    return self.eio.handle_request(environ, start_response)
  File "/var/www/rest-api-v2/venv/lib/python3.5/site-packages/engineio/server.py", line 268, in handle_request
    self.disconnect(sid)
  File "/var/www/rest-api-v2/venv/lib/python3.5/site-packages/engineio/server.py", line 202, in disconnect
    self._get_socket(sid).close()
  File "/var/www/rest-api-v2/venv/lib/python3.5/site-packages/engineio/server.py", line 411, in _get_socket
    raise KeyError('Session is disconnected')
KeyError: 'Session is disconnected'
```

**Comment by rizkiaditya24** [comment]
After two days trying to figure out what happened. I found that the problem is on my ISP, seems like they block port 8000/tcp that I used for WebSocket connection. It also causes error above, it disconnects itself when the port is blocking.

for more information about this similar issue:
https://nolanlawson.com/2013/05/31/web-sockets-with-socket-io-node-js-and-nginx-port-80-considered-harmful/#comment-85425

https://stackoverflow.com/questions/41638552/websocket-returns-status-200-instead-of-101-flask-socket-io
